### ISSUE TITLE
Error: Timeout performing GET {key}, inst: 1, queue: 2, qu=0, qs=2, qc=0, wr=0/0, in=0/0

### ISSUE THREAD

**Issue by Jackie-Yu** [StackExchange/StackExchange.Redis#83]
We are using StackExchange.Redis in Azure Redis Cache, However it occasionally throws the exception - 'Timeout performing GET {key}, inst: 1, queue: 2, qu=0, qs=2, qc=0, wr=0/0, in=0/0'.
Is there any setting or solution to resolve the issues?

**Comment by sesispla** [comment]
@mgravell same problem as @hewittj but with MVC 5.2.3 + Castle.Windsor 3.3.0
 **Immediate timeout** while trying to read data from Redis. Writing data to Redis works properly.

**Comment by Dorokhov** [comment]
the same on environment: Azure Redis, SE.Redis v.1.0.316.0, WCF Service

System.TimeoutException: Timeout performing PING, inst: 1, mgr: Inactive, queue: 15, qu=0, qs=15, qc=0, wr=0/0, in=497/0
   at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in c:\TeamCity\buildAgent\work\58bc9a6df18a3782\StackExchange.Redis\StackExchange\Redis\ConnectionMultiplexer.cs:line 1789

just restart of worker role instance helps, any suggestions?

**Comment by paveliak** [comment]
I cannot find official #316 version but it would date back to June 2014. I would recommend trying version that has PRs 127 and 129 merged. Latest #414 definitely has them, previous #394 may have too.

**Comment by vsthakur78** [comment]
Hi @hewittj this seems to be an older thread however I am getting similar Timeout issue on connection multiplexer line 1922. which to me result of timeout on the monitor.wait call up there at line 1859
Here is my setup: 
SE-Redis-1.0.450, 
Microsoft.AspNet.Mvc-5.2.3-beta1     
& Ninject 3.2.2.0 targetting .net-4.5.

Is it that SE-Redis version that I am using still has issues with mvc or ninject? I have done some fixes to my redis wrapper but still want to verify if this issue exists.
thanks.

**Comment by akashani** [comment]
I'm seeing a very similar problem with 1.0.450:

TimeoutException: Timeout performing HGETALL mykey, inst: 8, queue: 6, qu=0, qs=6, qc=0, wr=0/0, in=400/0

Based on the descriptions of the fields above, it seems like our problem is that 400 bytes are awaiting a read, but there are no readers processing them. Not sure if those 400 bytes are our 6 responses and the reason why qs=6 or if in fact we haven't received a response. 

Redis seems otherwise responsive (e.g. using redis-cli) and slowlog returns nothing (default threshold of  10000 microseconds.)  This is a development box with Redis running locally, so I can rule out any sort of network issues.

The only resolution seems to be to restart the app. Once it's stuck, no requests to Redis are processed.

**Comment by oviwankenobi** [comment]
azure redis db

Timeout performing HGET mykey, inst: 1, mgr: ExecuteSelect, err: never, queue: 3, qu: 0, qs: 3, qc: 0, wr: 0, wq: 0, in: 0, ar: 0, IOCP: (Busy=1,Free=999,Min=8,Max=1000), WORKER: (Busy=0,Free=4095,Min=8,Max=4095), clientName: theallmightyclient

   at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in c:\TeamCity\buildAgent\work\3ae0647004edff78\StackExchange.Redis\StackExchange\Redis\ConnectionMultiplexer.cs:line 1927
   at StackExchange.Redis.RedisBase.ExecuteSync[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in c:\TeamCity\buildAgent\work\3ae0647004edff78\StackExchange.Redis\StackExchange\Redis\RedisBase.cs:line 80
   at StackExchange.Redis.RedisDatabase.HashGet(RedisKey key, RedisValue hashField, CommandFlags flags) in c:\TeamCity\buildAgent\work\3ae0647004edff78\StackExchange.Redis\StackExchange\Redis\RedisDatabase.cs:line 118

thank you

**Comment by OlegLopatento** [comment]
Similar problem
Azure redis cache
Stackexchange.Redis 1.0.450
After several such errors all azure role instances are blocked.
Timeout performing GET somekey, inst: 1, mgr: Inactive, queue: 3, qu: 0, qs: 3, qc: 0, wr: 0, wq: 0, in: 65536, ar: 0, IOCP: (Busy=4,Free=1996,Min=1,Max=2000), WORKER: (Busy=2,Free=32765,Min=100,Max=32767)\r\n   at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in c:\\TeamCity\\buildAgent\\work\\3ae0647004edff78\\StackExchange.Redis\\StackExchange\\Redis\\ConnectionMultiplexer.cs:line 1922\r\n   at StackExchange.Redis.RedisBase.ExecuteSync[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in c:\\TeamCity\\buildAgent\\work\\3ae0647004edff78\\StackExchange.Redis\\StackExchange\\Redis\\RedisBase.cs:line 80\r\n   at StackExchange.Redis.RedisDatabase.StringGet(RedisKey key, CommandFlags flags) in c:\\TeamCity\\buildAgent\\work\\3ae0647004edff78\\StackExchange.Redis\\StackExchange\\Redis\\RedisDatabase.cs:line 1431\r\n   at 

Thanks

**Comment by OlegLopatento** [comment]
In my case setting PreserveAsyncOrder to false helped
See <a href="http://stackoverflow.com/questions/30797716/deadlock-when-accessing-stackexchange-redis">this</a>

**Comment by CraigSelbert** [comment]
I am getting the following.

System.Web.HttpException (0x80004005): Error executing child request for handler 'System.Web.Mvc.HttpHandlerUtil+ServerExecuteHttpHandlerAsyncWrapper'. ---> System.TimeoutException: Timeout performing GET Sections, inst: 1, mgr: Inactive, queue: 2, qu: 0, qs: 2, qc: 0, wr: 0, wq: 0, in: 1573, ar: 0, IOCP: (Busy=0,Free=1000,Min=2,Max=1000), WORKER: (Busy=9,Free=32758,Min=2,Max=32767)

   at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server)

**Comment by lywyn** [comment]
Using a dedicated Azure Redis Cache Server and Cloud Service ASP.NET app which uses the Redis Sessions State nuget packages;

Microsoft.Web.RedisSessionStateProvider version="1.6.4"
StackExchange.Redis.StrongName version="1.0.450"

We seem to get them in groups...

2015-08-12 10:29:03.030	Timeout performing EVAL, inst: 19, mgr: Inactive, queue: 6, qu: 0, qs: 6, qc: 0, wr: 0, wq: 0, in: 4879, ar: 0, IOCP: (Busy=1,Free=999,Min=1,Max=1000), WORKER: (Busy=10,Free=32757,Min=1,Max=32767)
2015-08-12 10:29:03.030	Timeout performing EVAL, inst: 21, mgr: Inactive, queue: 6, qu: 0, qs: 6, qc: 0, wr: 0, wq: 0, in: 5017, ar: 0, IOCP: (Busy=1,Free=999,Min=1,Max=1000), WORKER: (Busy=13,Free=32754,Min=1,Max=32767)
2015-08-12 10:29:03.030	Timeout performing EVAL, inst: 21, mgr: Inactive, queue: 6, qu: 0, qs: 6, qc: 0, wr: 0, wq: 0, in: 5017, ar: 0, IOCP: (Busy=1,Free=999,Min=1,Max=1000), WORKER: (Busy=13,Free=32754,Min=1,Max=32767)
2015-08-12 10:29:03.030	Timeout performing EVAL, inst: 21, mgr: Inactive, queue: 6, qu: 0, qs: 6, qc: 0, wr: 0, wq: 0, in: 4932, ar: 0, IOCP: (Busy=1,Free=999,Min=1,Max=1000), WORKER: (Busy=13,Free=32754,Min=1,Max=32767)

2015-08-20 09:02:07.197	Timeout performing EVAL, inst: 10, mgr: Inactive, queue: 4, qu: 0, qs: 4, qc: 0, wr: 0, wq: 0, in: 0, ar: 0, IOCP: (Busy=1,Free=999,Min=1,Max=1000), WORKER: (Busy=8,Free=32759,Min=1,Max=32767)
2015-08-20 09:02:14.063	Timeout performing EVAL, inst: 2, mgr: Inactive, queue: 6, qu: 0, qs: 6, qc: 0, wr: 0, wq: 0, in: 5844, ar: 0, IOCP: (Busy=1,Free=999,Min=1,Max=1000), WORKER: (Busy=8,Free=32759,Min=1,Max=32767)
2015-08-20 09:02:14.063	Timeout performing EVAL, inst: 2, mgr: Inactive, queue: 6, qu: 0, qs: 6, qc: 0, wr: 0, wq: 0, in: 5844, ar: 0, IOCP: (Busy=1,Free=999,Min=1,Max=1000), WORKER: (Busy=7,Free=32760,Min=1,Max=32767)

Any advice?

**Comment by engineering87** [comment]
Hi, I have the same problem:

MSG=ReadCounter;Eccezione:System.TimeoutException: Timeout performing GET LiveUpdateService_counter, inst: 20, mgr: ProcessReadQueue, queue: 2, qu=0, qs=2, qc=0, wr=0/0, in=8192/1     at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in RedisClient\StackExchange.Redis-master\StackExchange.Redis\StackExchange\Redis\ConnectionMultiplexer.cs:line 1802     at StackExchange.Redis.RedisBase.ExecuteSync[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in StackExchange.Redis-master\StackExchange.Redis\StackExchange\Redis\RedisBase.cs:line 81     at StackExchange.Redis.RedisDatabase.StringGet(RedisKey key, CommandFlags flags) in StackExchange.Redis-master\StackExchange.Redis\StackExchange\Redis\RedisDatabase.cs:line 1408 

Any advice?

**Comment by maulikhdave** [comment]
----------------------------------------------Looking for Urgent Help -------------------------------------------------

i have integrated rediss session  in my asp.net project with below dll

1)Microsoft.Web.RedisSessionStateProvider.dll  version 1.6.5.291

2)StackExchange.Redis.StrongName.dll              version 1.0.481.0

for some time it works fine then suddenly it started giving error below

__________________________________________________________________________________________________________________

Error, TagType=InfoTagModule=APPERROR$$$CRITICAL - APPERROR - Application_Error event - ErrorDetails - ** ** Error Report : Error: Timeout performing EVAL, inst: 0, mgr: ExecuteSelect, err: never, queue: 49, qu: 48, qs: 1, qc: 0, wr: 1, wq: 1, in: 0, ar: 0, IOCP: (Busy=0,Free=1000,Min=36,Max=1000), WORKER:(Busy=52,Free=32715,Min=36,Max=32767), clientName: WIN-4QJG4A59IU3

Source: StackExchange.Redis.StrongName  Stack Trace: at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) at StackExchange.Redis.RedisBase.ExecuteSync[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) at StackExchange.Redis.RedisDatabase.ScriptEvaluate(String script, RedisKey[] keys, RedisValue[] values, CommandFlags flags) at Microsoft.Web.Redis.StackExchangeClientConnection.<>c__DisplayClass4.<Eval>b__3() at Microsoft.Web.Redis.StackExchangeClientConnection.RetryForScriptNotFound(Func`1 redisOperation) at Microsoft.Web.Redis.StackExchangeClientConnection.RetryLogic(Func`1 redisOperation) at Microsoft.Web.Redis.StackExchangeClientConnection.Eval(String script, String[] keyArgs, Object[] valueArgs) at Microsoft.Web.Redis.RedisConnectionWrapper.TryUpdateAndReleaseLockIfLockIdMatch(Object lockId, ISessionStateItemCollection data, Int32 sessionTimeout) at Microsoft.Web.Redis.RedisSessionStateProvider.SetAndReleaseItemExclusive(HttpContext context, String id, SessionStateStoreData item, Object lockId, Boolean newItem) at System.Web.SessionState.SessionStateModule.OnReleaseState(Object source, EventArgs eventArgs) at System.Web.HttpApplication.SyncEventExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute() at System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously)  UserSession Items:User Key: 0 Currency Key: 3 Basket Key: 1454212 Date: 11 December 2015 Time:

i m using redis server from AWS

looking for urgent help i m using redis session

**Comment by oviwankenobi** [comment]
we got rid of these problems by actually using a different library, and we are quite happy with it
https://servicestack.net/redis
disclaimer: I have no affiliation whatsoever with the org building it

**Comment by jooper** [comment]
Set the settings as shown below.
![image](https://cloud.githubusercontent.com/assets/8216337/17961910/d51f25a8-6ae0-11e6-84b8-4f563c5a8c83.png)

**Comment by programcsharp** [comment]
Looks like this issue was never solved, but this seems to be the most elucidating thing I've found: http://stackoverflow.com/questions/30797716/deadlock-when-accessing-stackexchange-redis

Doing this "solves" it for me:

```
ConnectionMultiplexer connection = ...;
connection.PreserveAsyncOrder = false;
```
... but I don't know why.

I'm seeing the exact same deadlock symptom as above -- in the error message, qs says stuff is sent, in says stuff is received, but it hangs. Only restarting the process will fix it. This is with entirely sync calls to SE Redis, no async at all.

**Comment by rakletadmin** [comment]
thanks @jooper that definitely helped us.

**Comment by NickCraver** [comment]
Closing this out as the issues above have largely been solved. But more so, for those that haven't: we've added a lot more debugging information to the messages in current versions of the library to greatly narrow for each situation where in the pipe the problem is occurring. As a result of issues like this it greatly helped us know what to dump to figure these things out. Thanks everyone here for adding info :)
